### DISCUSSION TITLE
getting I2S realtime volume data from  INMP411 microphone

### DISCUSSION THREAD

**Discussion by Rumidom** [micropython/micropython#15558] [General]
Hi I'm trying to get realtime data from a INMP411 I2S microphone, to make it trigger the start of a recording after a certain volume level I was able to run the examples from this repository: https://github.com/miketeachman/micropython-i2s-examples , however I was unable to convert  the data I got from I2S.readinto(buf) into a usable format, more specifically I'm trying to convert the bytebuffer into a list of ints that I can use to measure volume level. Just to try to make same sense of the data I tried printing the buffer on the console and converting it to a numpy array on my computer and plot it with matplotlib: 

```python
dt = np.dtype("uint16")
dt = dt.newbyteorder('>')
plt.plot(np.frombuffer(bytearray_sample, dtype=dt))
```
I got this from a sample with music playing near by:
![image](https://github.com/user-attachments/assets/76da551e-7fb1-4abd-8278-519065234a0c)

and this from recording silence:
![image](https://github.com/user-attachments/assets/91ebccac-7493-4b2d-9dba-bf5310217460)

not sure why it ramps like that, am I using the wrong data type?

**Comment by Rumidom** [comment]
I changed the byte order from ```python dt.newbyteorder('>')``` to ```python dt.newbyteorder('<')``` and ```python np.dtype("uint16") ``` to  ```python np.dtype("int16") ``` and got something that makes more sense, but its still unusable: 

music: 
![image](https://github.com/user-attachments/assets/d72ec4ee-de32-4d59-89dd-ba5560282ab3)

silence: 
![image](https://github.com/user-attachments/assets/18b74f11-1bcc-42c3-897e-de22bf063476)

**Reply by ricksorensen** [reply]
I have similar results from a INMP441- with the little dip at beginning and the low signal noise.  Using both RP2040 and ESP32C3.  

I used a fixed (purish) tone signal out a small speaker 2" from microphone- and if I cut the first chunk out and amplify the remainder I can clearly see the wave form, and the spectral analysis matches the expected frequency.   But the level is very very low as in your example.

**Reply by Rumidom** [reply]
level is low because I'm using my laptop speakers, it seems like this dip is some sort of initialization of the mic, if I call audio_in.readinto() a few times before using the buffer it goes away, thanks!

**Comment by ricksorensen** [comment]
try signed int16 ('h'?)
